### DISCUSSION TITLE
UEFI ESRT device GUIDS

### DISCUSSION THREAD

**Discussion by Matombo** [fwupd/fwupd#5934] [Q&A] [answered]
What are the GUIDs of UEFI ESRT devices based on? Aka how unique are they?

e.g. Device ID: a45df35ac0e948ee180fe216a5f703f32dda163f in these dumps (one dump with flashrom plugin quirk and one without, note that the flashrom plugin seems to not remove the original esrt device, just changes the hirarchy a bit and creates a new device)

[fwupd_get-devices_no_quirk.txt](https://github.com/fwupd/fwupd/files/11850270/fwupd_get-devices_no_quirk.txt)

[fwupd_get-devices_quirk.txt](https://github.com/fwupd/fwupd/files/11850271/fwupd_get-devices_quirk.txt)

**Comment by hughsie** [comment, marked as answer]
> What are the GUIDs of UEFI ESRT devices based on? Aka how unique are they?

They're parsed from the ESRT table -- the system firmware has a combination of a hardcoded list of things in the ESRT and a boot-time hook to add other (mostly PCI) devices. The GUID is supposed to be unique for the model, i.e. so a different firmware requires a different GUID. The ESRT-based devices all use the capsule update mechanism.

The flashrom plugin doesn't have the luxury of using a hardware-provided table, so it uses some of the system HWIDs -- e.g. see https://github.com/fwupd/fwupd/tree/main/plugins/flashrom#guid-generation -- the hardware IDs themselves are shown in `sudo fwupdtool hwids` -- which tries to explain what DMI keys make up each GUID.

Slightly confusingly, hardware IDs are also used in what Microsoft call CHIDs. So you can target a firmware at ESRT GUID `FOO` and add the requirement that one of the CHIDs is **ALSO** GUID `BAR`. This solves the case where a vendor might have *forgotten* to change the ESRT GUID for a machine that actually needs different firmware.

Does that help?

**Reply by Matombo** [reply]
Thanks this helps.
